### DISCUSSION TITLE
UX - Updated play at speed

### DISCUSSION THREAD

**Discussion by JamesCrook** [audacity/audacity#876] [Ideas & Feature Requests]
This is an update to play at speed,  to allow negative speeds and looping.

![PlayAtSpeed](https://user-images.githubusercontent.com/1621525/117813786-6cad5900-b25b-11eb-83bd-4361e1e00bd9.gif)

_The icon for the play button changes in line with the speed (and direction).  (here x2, x3, x0.5, x-0.5 x-1)_  

* Looping can be turned on or off whilst playing
* Loop boundaries can be adjusted whilst playing, using the normal selection tool
* The loop is 'healed' (using repair effect on a few samples) so that it does not click.
* Speed can be adjusted whilst looping

Use the normal play button on the transport toolbar for initial speed of x1, the play button shown here for most recently used speed.  You can click the x1 speed button during a play at speed to switch back to x1 (a bit like A/B playing) or the play at speed button during a x1 speed play to reselect x2, x0.5 or other preset.

I would like to add 'drop down' options, on the play button, so that you can play-at-speed preserving or not preserving pitch, healing on/off etc.

**Comment by stmueller** [comment]
Some comments:
-Overall the concept looks great. integrating loop/speed makes a lot of sense.

-- Note: I implemented this original functionality in early 2000s, although it has been improved by others. There were several reasons why this was made a separate play button--partly because it was on a toolbar that was intended for a very narrow use case, and it did not make sense to impose that on all users, so I intentionally did not touch the main play button. I never considered it would be used broadly or enabled by default.  Because the other functions on the toolbar were removed and the toolbar  was enabled by default, it seems like it has become more useful to a larger group of users.  It does make sense to consider whether this should be a separate play button or if the functionality should just be integrated into the main play button.  It looks like this proposal is leaving it as a second play button.  I'm just pointing out the reasons it is a separate play button are not incredibly principled from an interaction-design perspective and it makes sense to weigh the option of having a single playback button.

-Merging into a single play button may carry with it confusions and the ability to set an alternative speed mode that gets remembered by the interface.  I would think podcast editors might use it that way--listen at double-speed until you get to a part that needs editing, then use the main play button to isolate what you need to edit, then return to double-speed playback.  Confusions because the playback speed and looping is only cosmetic and does not impact the saved file.

- I'm not sure there is a broad use case for playing in reverse, and especially loop-speed-shift-in-reverse. Actually, I don't even really understand how people use the loop function at all, but maybe it is as a preview for trimming and then exporting into a more sophisticated music system or prior to making a looped track. Maybe speeded playback in reverse is something those same kind of users would love to be able to do.  This would also suggest not 'healing' , because those users might be using the loop function specifically to identify a clean loop that won't click when exported or repeated on a separate track.  
- Is there a way to render or copy-at-speed to a new track?  Along with your other options proposed, this would essentially organize/replace about 4 plugins (speed, timeshift, repeat, reverse, etc.) and make them more directly accessible visually.  This seems like it might be solved if non-destructive effects are implemented.
- How would you handle a speed of 0? If someone could set it to 0 (a natural default value to accidentally reset it to after playing around with it)  it would be like a hidden pause function that makes it difficult to figure out why playback isn't working. 
- An alternate reverse would be as a modifier on the play button or toggle button like the loop toggle, rather than as a slider drag to negative values.  This would avoid the speed 0 problem.
- One problem with current speed slider is the difficulty in resetting to 1.0 default speed.  This does not seem to address that issue either, but a guide line at 1.0 that snaps-to would be useful  (not sure if this is possible).

Anyway, just some thoughts, and this would be great to see.

**Reply by JamesCrook** [reply]
About the original transcription toolbar - I would like to see it come back, and with a more general role, so that it can be a navigator between zones, but that's for another idea/discussion set.

> How would you handle a speed of 0? 

Right now, speed of 0 or equally very low speed indeed, is in effect a pause.  That's OK because the main play button on the main play/record toolbar always stays at x1 - so you won't lose the ability to play.   Secondly 90% of users who do set speed 0 will have done so whilst actually playing audio, so they will have seen and heard the audio slow down to zero, and know exactly what is going on.  This is in fact an argument for keeping the large x1 play button and the play at speed button separate.

Probably worthwhile, and would also avoid users getting stuck at zero speed and wondering why, is to always change the play-at-speed button back to a default of x2 when play is stopped.  It's simple and makes sure the two play buttons are visually different as well as actually doing different things.  That approach needs a bit more UX thinking... 

> I'm not sure there is a broad use case for playing in reverse

So this is about finding your place in audio and is related to scrubbing and seeking.  It can be useful to find the exact location of a click.  Because you continue to be able to adjust the clip boundaries as you play you get a kind of scrubbing too.  So you end up with two ways to locate a sound whilst listening, one that does not involve skipping over audio (using play at speed and positive and negative speeds) and one that does involve skipping over audio (using changes in selected region, as you play).

**Comment by petersampsonaudacity** [comment]
The ability to have negative Play at Speed would be a great asset.
